// devexpress-richedit/lib/rtf/import/dx-decoding/specified-symbols.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var map_creator_1 = require("../../../base-utils/map-creator");
exports.windows1250specifiedSymbols = new map_creator_1.MapCreator()
    .add(0x0080, 0x20AC)
    .add(0x0082, 0x201A)
    .add(0x0084, 0x201E)
    .add(0x0085, 0x2026)
    .add(0x0086, 0x2020)
    .add(0x0087, 0x2021)
    .add(0x0089, 0x2030)
    .add(0x008A, 0x0160)
    .add(0x008B, 0x2039)
    .add(0x008C, 0x015A)
    .add(0x008D, 0x0164)
    .add(0x008E, 0x017D)
    .add(0x008F, 0x0179)
    .add(0x0091, 0x2018)
    .add(0x0092, 0x2019)
    .add(0x0093, 0x201C)
    .add(0x0094, 0x201D)
    .add(0x0095, 0x2022)
    .add(0x0096, 0x2013)
    .add(0x0097, 0x2014)
    .add(0x0099, 0x2122)
    .add(0x009A, 0x0161)
    .add(0x009B, 0x203A)
    .add(0x009C, 0x015B)
    .add(0x009D, 0x0165)
    .add(0x009E, 0x017E)
    .add(0x009F, 0x017A)
    .add(0x00A0, 0x00A0)
    .add(0x00A1, 0x02C7)
    .add(0x00A2, 0x02D8)
    .add(0x00A3, 0x0141)
    .add(0x00A4, 0x00A4)
    .add(0x00A5, 0x0104)
    .add(0x00A6, 0x00A6)
    .add(0x00A7, 0x00A7)
    .add(0x00A8, 0x00A8)
    .add(0x00A9, 0x00A9)
    .add(0x00AA, 0x015E)
    .add(0x00AB, 0x00AB)
    .add(0x00AC, 0x00AC)
    .add(0x00AD, 0x00AD)
    .add(0x00AE, 0x00AE)
    .add(0x00AF, 0x017B)
    .add(0x00B0, 0x00B0)
    .add(0x00B1, 0x00B1)
    .add(0x00B2, 0x02DB)
    .add(0x00B3, 0x0142)
    .add(0x00B4, 0x00B4)
    .add(0x00B5, 0x00B5)
    .add(0x00B6, 0x00B6)
    .add(0x00B7, 0x00B7)
    .add(0x00B8, 0x00B8)
    .add(0x00B9, 0x0105)
    .add(0x00BA, 0x015F)
    .add(0x00BB, 0x00BB)
    .add(0x00BC, 0x013D)
    .add(0x00BD, 0x02DD)
    .add(0x00BE, 0x013E)
    .add(0x00BF, 0x017C)
    .add(0x00C0, 0x0154)
    .add(0x00C1, 0x00C1)
    .add(0x00C2, 0x00C2)
    .add(0x00C3, 0x0102)
    .add(0x00C4, 0x00C4)
    .add(0x00C5, 0x0139)
    .add(0x00C6, 0x0106)
    .add(0x00C7, 0x00C7)
    .add(0x00C8, 0x010C)
    .add(0x00C9, 0x00C9)
    .add(0x00CA, 0x0118)
    .add(0x00CB, 0x00CB)
    .add(0x00CC, 0x011A)
    .add(0x00CD, 0x00CD)
    .add(0x00CE, 0x00CE)
    .add(0x00CF, 0x010E)
    .add(0x00D0, 0x0110)
    .add(0x00D1, 0x0143)
    .add(0x00D2, 0x0147)
    .add(0x00D3, 0x00D3)
    .add(0x00D4, 0x00D4)
    .add(0x00D5, 0x0150)
    .add(0x00D6, 0x00D6)
    .add(0x00D7, 0x00D7)
    .add(0x00D8, 0x0158)
    .add(0x00D9, 0x016E)
    .add(0x00DA, 0x00DA)
    .add(0x00DB, 0x0170)
    .add(0x00DC, 0x00DC)
    .add(0x00DD, 0x00DD)
    .add(0x00DE, 0x0162)
    .add(0x00DF, 0x00DF)
    .add(0x00E0, 0x0155)
    .add(0x00E1, 0x00E1)
    .add(0x00E2, 0x00E2)
    .add(0x00E3, 0x0103)
    .add(0x00E4, 0x00E4)
    .add(0x00E5, 0x013A)
    .add(0x00E6, 0x0107)
    .add(0x00E7, 0x00E7)
    .add(0x00E8, 0x010D)
    .add(0x00E9, 0x00E9)
    .add(0x00EA, 0x0119)
    .add(0x00EB, 0x00EB)
    .add(0x00EC, 0x011B)
    .add(0x00ED, 0x00ED)
    .add(0x00EE, 0x00EE)
    .add(0x00EF, 0x010F)
    .add(0x00F0, 0x0111)
    .add(0x00F1, 0x0144)
    .add(0x00F2, 0x0148)
    .add(0x00F3, 0x00F3)
    .add(0x00F4, 0x00F4)
    .add(0x00F5, 0x0151)
    .add(0x00F6, 0x00F6)
    .add(0x00F7, 0x00F7)
    .add(0x00F8, 0x0159)
    .add(0x00F9, 0x016F)
    .add(0x00FA, 0x00FA)
    .add(0x00FB, 0x0171)
    .add(0x00FC, 0x00FC)
    .add(0x00FD, 0x00FD)
    .add(0x00FE, 0x0163)
    .add(0x00FF, 0x02D9)
    .get();
exports.windows1252specifiedSymbols = new map_creator_1.MapCreator()
    .add(0x0080, 0x20AC)
    .add(0x0082, 0x201A)
    .add(0x0083, 0x0192)
    .add(0x0084, 0x201E)
    .add(0x0085, 0x2026)
    .add(0x0086, 0x2020)
    .add(0x0087, 0x2021)
    .add(0x0088, 0x02C6)
    .add(0x0089, 0x2030)
    .add(0x008A, 0x0160)
    .add(0x008B, 0x2039)
    .add(0x008C, 0x0152)
    .add(0x008E, 0x017D)
    .add(0x0091, 0x2018)
    .add(0x0092, 0x2019)
    .add(0x0093, 0x201C)
    .add(0x0094, 0x201D)
    .add(0x0095, 0x2022)
    .add(0x0096, 0x2013)
    .add(0x0097, 0x2014)
    .add(0x0098, 0x02DC)
    .add(0x0099, 0x2122)
    .add(0x009A, 0x0161)
    .add(0x009B, 0x203A)
    .add(0x009C, 0x0153)
    .add(0x009E, 0x017E)
    .add(0x009F, 0x0178)
    .get();
exports.windows1253specifiedSymbols = new map_creator_1.MapCreator()
    .add(0x0080, 0x20AC)
    .add(0x0082, 0x201A)
    .add(0x0083, 0x0192)
    .add(0x0084, 0x201E)
    .add(0x0085, 0x2026)
    .add(0x0086, 0x2020)
    .add(0x0087, 0x2021)
    .add(0x0089, 0x2030)
    .add(0x008B, 0x2039)
    .add(0x0091, 0x2018)
    .add(0x0092, 0x2019)
    .add(0x0093, 0x201C)
    .add(0x0094, 0x201D)
    .add(0x0095, 0x2022)
    .add(0x0096, 0x2013)
    .add(0x0097, 0x2014)
    .add(0x0099, 0x2122)
    .add(0x009B, 0x203A)
    .add(0x00A1, 0x0385)
    .add(0x00A2, 0x0386)
    .add(0x00AF, 0x2015)
    .add(0x00B4, 0x0384)
    .add(0x00B8, 0x0388)
    .add(0x00B9, 0x0389)
    .add(0x00BA, 0x038A)
    .add(0x00BC, 0x038C)
    .add(0x00BE, 0x038E)
    .add(0x00BF, 0x038F)
    .add(0x00C0, 0x0390)
    .add(0x00C1, 0x0391)
    .add(0x00C2, 0x0392)
    .add(0x00C3, 0x0393)
    .add(0x00C4, 0x0394)
    .add(0x00C5, 0x0395)
    .add(0x00C6, 0x0396)
    .add(0x00C7, 0x0397)
    .add(0x00C8, 0x0398)
    .add(0x00C9, 0x0399)
    .add(0x00CA, 0x039A)
    .add(0x00CB, 0x039B)
    .add(0x00CC, 0x039C)
    .add(0x00CD, 0x039D)
    .add(0x00CE, 0x039E)
    .add(0x00CF, 0x039F)
    .add(0x00D0, 0x03A0)
    .add(0x00D1, 0x03A1)
    .add(0x00D3, 0x03A3)
    .add(0x00D4, 0x03A4)
    .add(0x00D5, 0x03A5)
    .add(0x00D6, 0x03A6)
    .add(0x00D7, 0x03A7)
    .add(0x00D8, 0x03A8)
    .add(0x00D9, 0x03A9)
    .add(0x00DA, 0x03AA)
    .add(0x00DB, 0x03AB)
    .add(0x00DC, 0x03AC)
    .add(0x00DD, 0x03AD)
    .add(0x00DE, 0x03AE)
    .add(0x00DF, 0x03AF)
    .add(0x00E0, 0x03B0)
    .add(0x00E1, 0x03B1)
    .add(0x00E2, 0x03B2)
    .add(0x00E3, 0x03B3)
    .add(0x00E4, 0x03B4)
    .add(0x00E5, 0x03B5)
    .add(0x00E6, 0x03B6)
    .add(0x00E7, 0x03B7)
    .add(0x00E8, 0x03B8)
    .add(0x00E9, 0x03B9)
    .add(0x00EA, 0x03BA)
    .add(0x00EB, 0x03BB)
    .add(0x00EC, 0x03BC)
    .add(0x00ED, 0x03BD)
    .add(0x00EE, 0x03BE)
    .add(0x00EF, 0x03BF)
    .add(0x00F0, 0x03C0)
    .add(0x00F1, 0x03C1)
    .add(0x00F2, 0x03C2)
    .add(0x00F3, 0x03C3)
    .add(0x00F4, 0x03C4)
    .add(0x00F5, 0x03C5)
    .add(0x00F6, 0x03C6)
    .add(0x00F7, 0x03C7)
    .add(0x00F8, 0x03C8)
    .add(0x00F9, 0x03C9)
    .add(0x00FA, 0x03CA)
    .add(0x00FB, 0x03CB)
    .add(0x00FC, 0x03CC)
    .add(0x00FD, 0x03CD)
    .add(0x00FE, 0x03CE)
    .get();
exports.windows1254specifiedSymbols = new map_creator_1.MapCreator()
    .add(0x0080, 0x20AC)
    .add(0x0082, 0x201A)
    .add(0x0083, 0x0192)
    .add(0x0084, 0x201E)
    .add(0x0085, 0x2026)
    .add(0x0086, 0x2020)
    .add(0x0087, 0x2021)
    .add(0x0088, 0x02C6)
    .add(0x0089, 0x2030)
    .add(0x008A, 0x0160)
    .add(0x008B, 0x2039)
    .add(0x008C, 0x0152)
    .add(0x0091, 0x2018)
    .add(0x0092, 0x2019)
    .add(0x0093, 0x201C)
    .add(0x0094, 0x201D)
    .add(0x0095, 0x2022)
    .add(0x0096, 0x2013)
    .add(0x0097, 0x2014)
    .add(0x0098, 0x02DC)
    .add(0x0099, 0x2122)
    .add(0x009A, 0x0161)
    .add(0x009B, 0x203A)
    .add(0x009C, 0x0153)
    .add(0x009F, 0x0178)
    .add(0x00D0, 0x011E)
    .add(0x00DD, 0x0130)
    .add(0x00DE, 0x015E)
    .add(0x00F0, 0x011F)
    .add(0x00FD, 0x0131)
    .add(0x00FE, 0x015F)
    .get();
exports.windows1255specifiedSymbols = new map_creator_1.MapCreator()
    .add(0x0080, 0x20AC)
    .add(0x0082, 0x201A)
    .add(0x0083, 0x0192)
    .add(0x0084, 0x201E)
    .add(0x0085, 0x2026)
    .add(0x0086, 0x2020)
    .add(0x0087, 0x2021)
    .add(0x0088, 0x02C6)
    .add(0x0089, 0x2030)
    .add(0x008B, 0x2039)
    .add(0x0091, 0x2018)
    .add(0x0092, 0x2019)
    .add(0x0093, 0x201C)
    .add(0x0094, 0x201D)
    .add(0x0095, 0x2022)
    .add(0x0096, 0x2013)
    .add(0x0097, 0x2014)
    .add(0x0098, 0x02DC)
    .add(0x0099, 0x2122)
    .add(0x009B, 0x203A)
    .add(0x00A4, 0x20AA)
    .add(0x00AA, 0x00D7)
    .add(0x00BA, 0x00F7)
    .add(0x00C0, 0x05B0)
    .add(0x00C1, 0x05B1)
    .add(0x00C2, 0x05B2)
    .add(0x00C3, 0x05B3)
    .add(0x00C4, 0x05B4)
    .add(0x00C5, 0x05B5)
    .add(0x00C6, 0x05B6)
    .add(0x00C7, 0x05B7)
    .add(0x00C8, 0x05B8)
    .add(0x00C9, 0x05B9)
    .add(0x00CA, 0x05BA)
    .add(0x00CB, 0x05BB)
    .add(0x00CC, 0x05BC)
    .add(0x00CD, 0x05BD)
    .add(0x00CE, 0x05BE)
    .add(0x00CF, 0x05BF)
    .add(0x00D0, 0x05C0)
    .add(0x00D1, 0x05C1)
    .add(0x00D2, 0x05C2)
    .add(0x00D3, 0x05C3)
    .add(0x00D4, 0x05F0)
    .add(0x00D5, 0x05F1)
    .add(0x00D6, 0x05F2)
    .add(0x00D7, 0x05F3)
    .add(0x00D8, 0x05F4)
    .add(0x00E0, 0x05D0)
    .add(0x00E1, 0x05D1)
    .add(0x00E2, 0x05D2)
    .add(0x00E3, 0x05D3)
    .add(0x00E4, 0x05D4)
    .add(0x00E5, 0x05D5)
    .add(0x00E6, 0x05D6)
    .add(0x00E7, 0x05D7)
    .add(0x00E8, 0x05D8)
    .add(0x00E9, 0x05D9)
    .add(0x00EA, 0x05DA)
    .add(0x00EB, 0x05DB)
    .add(0x00EC, 0x05DC)
    .add(0x00ED, 0x05DD)
    .add(0x00EE, 0x05DE)
    .add(0x00EF, 0x05DF)
    .add(0x00F0, 0x05E0)
    .add(0x00F1, 0x05E1)
    .add(0x00F2, 0x05E2)
    .add(0x00F3, 0x05E3)
    .add(0x00F4, 0x05E4)
    .add(0x00F5, 0x05E5)
    .add(0x00F6, 0x05E6)
    .add(0x00F7, 0x05E7)
    .add(0x00F8, 0x05E8)
    .add(0x00F9, 0x05E9)
    .add(0x00FA, 0x05EA)
    .add(0x00FD, 0x200E)
    .add(0x00FE, 0x200F)
    .get();
exports.windows1256specifiedSymbols = new map_creator_1.MapCreator()
    .add(0x0080, 0x20AC)
    .add(0x0081, 0x067E)
    .add(0x0082, 0x201A)
    .add(0x0083, 0x0192)
    .add(0x0084, 0x201E)
    .add(0x0085, 0x2026)
    .add(0x0086, 0x2020)
    .add(0x0087, 0x2021)
    .add(0x0088, 0x02C6)
    .add(0x0089, 0x2030)
    .add(0x008A, 0x0679)
    .add(0x008B, 0x2039)
    .add(0x008C, 0x0152)
    .add(0x008D, 0x0686)
    .add(0x008E, 0x0698)
    .add(0x008F, 0x0688)
    .add(0x0090, 0x06AF)
    .add(0x0091, 0x2018)
    .add(0x0092, 0x2019)
    .add(0x0093, 0x201C)
    .add(0x0094, 0x201D)
    .add(0x0095, 0x2022)
    .add(0x0096, 0x2013)
    .add(0x0097, 0x2014)
    .add(0x0098, 0x06A9)
    .add(0x0099, 0x2122)
    .add(0x009A, 0x0691)
    .add(0x009B, 0x203A)
    .add(0x009C, 0x0153)
    .add(0x009D, 0x200C)
    .add(0x009E, 0x200D)
    .add(0x009F, 0x06BA)
    .add(0x00A1, 0x060C)
    .add(0x00AA, 0x06BE)
    .add(0x00BA, 0x061B)
    .add(0x00BF, 0x061F)
    .add(0x00C0, 0x06C1)
    .add(0x00C1, 0x0621)
    .add(0x00C2, 0x0622)
    .add(0x00C3, 0x0623)
    .add(0x00C4, 0x0624)
    .add(0x00C5, 0x0625)
    .add(0x00C6, 0x0626)
    .add(0x00C7, 0x0627)
    .add(0x00C8, 0x0628)
    .add(0x00C9, 0x0629)
    .add(0x00CA, 0x062A)
    .add(0x00CB, 0x062B)
    .add(0x00CC, 0x062C)
    .add(0x00CD, 0x062D)
    .add(0x00CE, 0x062E)
    .add(0x00CF, 0x062F)
    .add(0x00D0, 0x0630)
    .add(0x00D1, 0x0631)
    .add(0x00D2, 0x0632)
    .add(0x00D3, 0x0633)
    .add(0x00D4, 0x0634)
    .add(0x00D5, 0x0635)
    .add(0x00D6, 0x0636)
    .add(0x00D8, 0x0637)
    .add(0x00D9, 0x0638)
    .add(0x00DA, 0x0639)
    .add(0x00DB, 0x063A)
    .add(0x00DC, 0x0640)
    .add(0x00DD, 0x0641)
    .add(0x00DE, 0x0642)
    .add(0x00DF, 0x0643)
    .add(0x00E1, 0x0644)
    .add(0x00E3, 0x0645)
    .add(0x00E4, 0x0646)
    .add(0x00E5, 0x0647)
    .add(0x00E6, 0x0648)
    .add(0x00EC, 0x0649)
    .add(0x00ED, 0x064A)
    .add(0x00F0, 0x064B)
    .add(0x00F1, 0x064C)
    .add(0x00F2, 0x064D)
    .add(0x00F3, 0x064E)
    .add(0x00F5, 0x064F)
    .add(0x00F6, 0x0650)
    .add(0x00F8, 0x0651)
    .add(0x00FA, 0x0652)
    .add(0x00FD, 0x200E)
    .add(0x00FE, 0x200F)
    .add(0x00FF, 0x06D2)
    .get();
exports.windows1257specifiedSymbols = new map_creator_1.MapCreator()
    .add(0x0080, 0x20AC)
    .add(0x0082, 0x201A)
    .add(0x0084, 0x201E)
    .add(0x0085, 0x2026)
    .add(0x0086, 0x2020)
    .add(0x0087, 0x2021)
    .add(0x0089, 0x2030)
    .add(0x008B, 0x2039)
    .add(0x008D, 0x00A8)
    .add(0x008E, 0x02C7)
    .add(0x008F, 0x00B8)
    .add(0x009C, 0x00AF)
    .add(0x009D, 0x02DB)
    .get();
exports.windows1258specifiedSymbols = new map_creator_1.MapCreator()
    .add(0x0080, 0x20AC)
    .add(0x0082, 0x201A)
    .add(0x0084, 0x201E)
    .add(0x0085, 0x2026)
    .add(0x0086, 0x2020)
    .add(0x0087, 0x2021)
    .add(0x0089, 0x2030)
    .add(0x008B, 0x2039)
    .add(0x008D, 0x00A8)
    .add(0x008E, 0x02C7)
    .add(0x008F, 0x00B8)
    .add(0x009C, 0x00AF)
    .add(0x009D, 0x02DB)
    .get();
