// devexpress-richedit/lib/base/commands/floating-objects/floating-object-drag-drop-change-position-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var layout_position_creator_1 = require("../../../core/layout-engine/layout-position-creator");
var document_layout_details_level_1 = require("../../../core/layout/document-layout-details-level");
var layout_point_1 = require("../../../core/layout/layout-point");
var field_1 = require("../../../core/model/fields/field");
var enums_1 = require("../../../core/model/floating-objects/enums");
var insert_picture_manipulator_params_1 = require("../../../core/model/manipulators/picture-manipulator/insert-picture-manipulator-params");
var image_loading_options_1 = require("../../../core/model/manipulators/picture-manipulator/loader/image-loading-options");
var text_box_manipulator_1 = require("../../../core/model/manipulators/text-box-manipulator");
var run_type_1 = require("../../../core/model/runs/run-type");
var sub_document_1 = require("../../../core/model/sub-document");
var unit_converter_1 = require("@devexpress/utils/lib/class/unit-converter");
var point_1 = require("@devexpress/utils/lib/geometry/point");
var fixed_1 = require("@devexpress/utils/lib/intervals/fixed");
var selection_history_item_1 = require("../../model/history/selection/selection-history-item");
var model_states_1 = require("../../scroll/model-states");
var command_base_1 = require("../command-base");
var command_states_1 = require("../command-states");
var FloatingObjectMovedArgumentInner = (function () {
    function FloatingObjectMovedArgumentInner(subDocumentId, newPosition, pageIntervals, pageIndex, objectX, objectY) {
        this.newPosition = newPosition;
        this.subDocumentId = subDocumentId;
        this.pageIntervals = pageIntervals;
        this.pageIndex = pageIndex;
        this.objectX = objectX;
        this.objectY = objectY;
        if (this.pageIndex == 0 && this.pageIntervals[0].start != 0)
            this.pageIntervals.splice(0, 1, new fixed_1.FixedInterval(0, this.pageIntervals[0].end));
    }
    return FloatingObjectMovedArgumentInner;
}());
exports.FloatingObjectMovedArgumentInner = FloatingObjectMovedArgumentInner;
var FloatingObjectDragDropChangePositionCommand = (function (_super) {
    tslib_1.__extends(FloatingObjectDragDropChangePositionCommand, _super);
    function FloatingObjectDragDropChangePositionCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FloatingObjectDragDropChangePositionCommand.prototype.getState = function () {
        return new command_states_1.SimpleCommandState(this.isEnabled());
    };
    FloatingObjectDragDropChangePositionCommand.prototype.canModify = function () {
        return true;
    };
    FloatingObjectDragDropChangePositionCommand.prototype.isEnabled = function () {
        var specialRunInfo = this.selection.specialRunInfo;
        return _super.prototype.isEnabled.call(this) && specialRunInfo.isSelected() && specialRunInfo.isSelectedAnchorObject &&
            this.getFloatingObjectParentSubDocument().isEditable([new fixed_1.FixedInterval(specialRunInfo.getPosition(), 1)]);
    };
    FloatingObjectDragDropChangePositionCommand.prototype.executeCore = function (_state, options) {
        var topInfo = this.control.viewManager.canvasManager.getScrollTopInfo();
        var subDocument = this.selection.activeSubDocument;
        var runInfo = subDocument.getRunAndIndexesByPosition(this.selection.specialRunInfo.getPosition());
        var oldRun = runInfo.run.clone();
        var oldRunPos = runInfo.getAbsoluteRunPosition();
        var finalClickPoint = options.finalClickPoint;
        var initialHtr = this.control.hitTestManager.calculate(new layout_point_1.LayoutPoint(options.endPageIndex, finalClickPoint.x, finalClickPoint.y), document_layout_details_level_1.DocumentLayoutDetailsLevel.Character, subDocument);
        initialHtr.correctAsVisibleBox();
        this.history.beginTransaction();
        this.move(subDocument, oldRun, oldRunPos, initialHtr, options.endPageIndex, options.finalPoint);
        this.history.endTransaction();
        this.selection.scrollManager.setScroll(new model_states_1.ScrollState().byScrollInfo.setPageInfo(topInfo));
        return true;
    };
    FloatingObjectDragDropChangePositionCommand.prototype.move = function (subDocument, oldRun, oldRunPos, initialHtr, endPageIndex, finalPoint) {
        if (initialHtr.row.tableCellInfo) {
            var paragraphStartHTR = this.findParagraphStartOnThisPage(initialHtr);
            var newLogPos = this.getNewLogPosition(subDocument, paragraphStartHTR);
            var arg = new FloatingObjectMovedArgumentInner(subDocument.id, newLogPos, paragraphStartHTR.page.getContentIntervals(), paragraphStartHTR.pageIndex, finalPoint.x, finalPoint.y);
            this.control.owner.raiseFloatingObjectMovedObject(arg);
            if (newLogPos != arg.newPosition || finalPoint.x != arg.objectX || finalPoint.y != arg.objectY) {
                this.handleEvent(arg, subDocument, oldRun, oldRunPos, false);
            }
            else {
                this.moveInsideTable(oldRun, oldRunPos, newLogPos, paragraphStartHTR, finalPoint, false);
            }
        }
        else {
            this.moveOutsideTable(subDocument, oldRun, oldRunPos, endPageIndex, finalPoint);
        }
    };
    FloatingObjectDragDropChangePositionCommand.prototype.handleEvent = function (arg, subDocument, oldRun, oldRunPos, oldIntervalRemoved) {
        var lp = subDocument.isMain() ?
            layout_position_creator_1.LayoutPositionMainSubDocumentCreator.ensureLayoutPosition(this.control.layoutFormatterManager, subDocument, arg.newPosition, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character, new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false)) :
            new layout_position_creator_1.LayoutPositionOtherSubDocumentCreator(this.control.layout, subDocument, arg.newPosition, this.control.selection.pageIndex, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character)
                .create(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(false), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true));
        var finalPos = new point_1.Point(arg.objectX, arg.objectY);
        if (lp.row.tableCellInfo)
            this.moveInsideTable(oldRun, oldRunPos, arg.newPosition, lp, finalPos, oldIntervalRemoved);
        else {
            var newPoint = new point_1.Point(finalPos.x - lp.getLayoutX(null, document_layout_details_level_1.DocumentLayoutDetailsLevel.Column), finalPos.y);
            var newAnchorInfo = this.getNewAnchorInfo(oldRun.anchorInfo, newPoint);
            if (!oldIntervalRemoved)
                this.modelManipulator.range.removeInterval(this.selection.intervalsInfo.subDocInterval, false, false);
            this.addRun(oldRun, arg.newPosition, newAnchorInfo);
            this.history.addAndRedo(new selection_history_item_1.SelectionHistoryItem(this.modelManipulator, this.selection, this.selection.getState(), this.selection.getState().setInterval(new fixed_1.FixedInterval(arg.newPosition, 1))));
        }
    };
    FloatingObjectDragDropChangePositionCommand.prototype.moveInsideTable = function (oldRun, oldRunPos, newLogPos, finalPosHTR, finalPoint, oldIntervalRemoved) {
        var offset = finalPosHTR.row.tableCellInfo ?
            new point_1.Point(finalPosHTR.row.tableCellInfo.x, finalPosHTR.row.tableCellInfo.y) :
            new point_1.Point(0, 0);
        var newPoint = new point_1.Point(finalPoint.x - finalPosHTR.getLayoutX(null, document_layout_details_level_1.DocumentLayoutDetailsLevel.Column) - offset.x, finalPoint.y - finalPosHTR.getLayoutY(document_layout_details_level_1.DocumentLayoutDetailsLevel.Column) - offset.y);
        if (!oldIntervalRemoved)
            this.modelManipulator.range.removeInterval(this.selection.intervalsInfo.subDocInterval, false, false);
        this.control.layoutFormatterManager.forceFormatPage(finalPosHTR.pageIndex);
        var newRunPos = newLogPos + (oldRunPos < newLogPos ? -1 : 0);
        this.addRun(oldRun, newRunPos, this.getNewAnchorInfo(oldRun.anchorInfo, newPoint));
        this.history.addAndRedo(new selection_history_item_1.SelectionHistoryItem(this.modelManipulator, this.selection, this.selection.getState(), this.selection.getState().setInterval(new fixed_1.FixedInterval(newRunPos, 1))));
    };
    FloatingObjectDragDropChangePositionCommand.prototype.moveOutsideTable = function (subDocument, oldRun, oldRunPos, endPageIndex, pagePosition) {
        this.modelManipulator.range.removeInterval(this.selection.intervalsInfo.subDocInterval, false, false);
        this.control.layoutFormatterManager.forceFormatPage(endPageIndex);
        var layoutPoint = new layout_point_1.LayoutPoint(Math.min(endPageIndex, this.control.layout.pages.length), pagePosition.x, pagePosition.y);
        var htr = this.control.hitTestManager.calculate(layoutPoint, document_layout_details_level_1.DocumentLayoutDetailsLevel.Row, subDocument);
        var paragraphStartHTR = this.findParagraphStartOnThisPage(htr);
        var newLogPos = this.getNewLogPosition(subDocument, paragraphStartHTR);
        var arg = new FloatingObjectMovedArgumentInner(subDocument.id, newLogPos, paragraphStartHTR.page.getContentIntervals(), paragraphStartHTR.pageIndex, pagePosition.x, pagePosition.y);
        this.control.owner.raiseFloatingObjectMovedObject(arg);
        if (newLogPos != arg.newPosition || pagePosition.x != arg.objectX || pagePosition.y != arg.objectY) {
            this.handleEvent(arg, subDocument, oldRun, oldRunPos, true);
        }
        else {
            var newPoint = new point_1.Point(pagePosition.x - htr.getLayoutX(null, document_layout_details_level_1.DocumentLayoutDetailsLevel.Column), pagePosition.y);
            var newAnchorInfo = this.getNewAnchorInfo(oldRun.anchorInfo, newPoint);
            this.addRun(oldRun, newLogPos, newAnchorInfo);
            this.history.addAndRedo(new selection_history_item_1.SelectionHistoryItem(this.modelManipulator, this.selection, this.selection.getState(), this.selection.getState().setInterval(new fixed_1.FixedInterval(newLogPos, 1))));
        }
    };
    FloatingObjectDragDropChangePositionCommand.prototype.findParagraphStartOnThisPage = function (htr) {
        var paragraph = this.selection.activeSubDocument.getParagraphByPosition(htr.getLogPosition());
        while (htr.stepBackRow()) {
            if (!paragraph.interval.contains(htr.getLogPosition())) {
                htr.stepForwardRow();
                break;
            }
        }
        return htr;
    };
    FloatingObjectDragDropChangePositionCommand.prototype.addRun = function (oldRun, position, anchorInfo) {
        if (oldRun.getType() == run_type_1.RunType.AnchoredPictureRun) {
            var pictureRun = oldRun;
            this.modelManipulator.picture.insertAnchoredPictureViaHistory(new sub_document_1.SubDocumentPosition(this.selection.activeSubDocument, position), this.inputPosition.charPropsBundle, new insert_picture_manipulator_params_1.AnchorPictureInfo(pictureRun.size.clone(), pictureRun.shape.clone(), anchorInfo, pictureRun.info.containerProperties.clone(), pictureRun.info.nonVisualDrawingProperties.clone()), new image_loading_options_1.ImageLoadingOptions(false));
        }
        else {
            var textBoxRun = oldRun;
            var originalSubDocument = this.control.modelManager.model.subDocuments[textBoxRun.subDocId];
            this.modelManipulator.textBox.insertAnchoredTextBoxViaHistoty(new sub_document_1.SubDocumentPosition(this.selection.activeSubDocument, position), this.inputPosition.charPropsBundle, new text_box_manipulator_1.BaseTextBoxInfo(null, textBoxRun.size.clone(), textBoxRun.shape.clone(), anchorInfo, textBoxRun.textBoxProperties.clone()));
            if (originalSubDocument.getDocumentEndPosition() > 1) {
                var newTextBoxRun = this.control.selection.activeSubDocument.getRunAndIndexesByPosition(position).run;
                this.control.modelManager.modelManipulator.subDocument.insertSubDocument(new sub_document_1.SubDocumentPosition(this.control.modelManager.model.subDocuments[newTextBoxRun.subDocId], 0), new sub_document_1.SubDocumentInterval(originalSubDocument, new fixed_1.FixedInterval(0, originalSubDocument.getDocumentEndPosition() - 1)));
            }
        }
    };
    FloatingObjectDragDropChangePositionCommand.prototype.getNewAnchorInfo = function (anchorInfo, newPoint) {
        anchorInfo = anchorInfo.clone();
        anchorInfo.horizontalPositionAlignment = enums_1.AnchorObjectHorizontalPositionAlignment.None;
        anchorInfo.verticalPositionAlignment = enums_1.AnchorObjectVerticalPositionAlignment.None;
        anchorInfo.horizontalPositionType = enums_1.AnchorObjectHorizontalPositionType.Column;
        anchorInfo.verticalPositionType = enums_1.AnchorObjectVerticalPositionType.Page;
        anchorInfo.percentOffset.x = 0;
        anchorInfo.percentOffset.y = 0;
        anchorInfo.offset.x = unit_converter_1.UnitConverter.pixelsToTwips(newPoint.x);
        anchorInfo.offset.y = unit_converter_1.UnitConverter.pixelsToTwips(newPoint.y);
        return anchorInfo;
    };
    FloatingObjectDragDropChangePositionCommand.prototype.getNewLogPosition = function (subDocument, htr) {
        var pos = htr.getLogPosition(document_layout_details_level_1.DocumentLayoutDetailsLevel.Row);
        if (!subDocument.fields.length)
            return pos;
        var index = Math.max(0, field_1.Field.normedBinaryIndexOf(subDocument.fields, pos));
        var topLevelField = subDocument.fields[index].getAbsolutelyTopLevelField();
        var startFieldPos = topLevelField.getFieldStartPosition();
        for (var ind = topLevelField.index, field = void 0; field = subDocument.fields[ind]; ind++) {
            if (field.getFieldStartPosition() >= topLevelField.getFieldEndPosition())
                break;
            if (pos == field.getResultStartPosition()) {
                var lp = subDocument.isMain() ?
                    layout_position_creator_1.LayoutPositionMainSubDocumentCreator.ensureLayoutPosition(this.control.layoutFormatterManager, subDocument, pos, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character, new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true)) :
                    new layout_position_creator_1.LayoutPositionOtherSubDocumentCreator(this.control.layout, subDocument, pos, this.control.selection.pageIndex, document_layout_details_level_1.DocumentLayoutDetailsLevel.Character)
                        .create(new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true), new layout_position_creator_1.LayoutPositionCreatorConflictFlags().setDefault(true));
                if (startFieldPos >= lp.getLogPosition(document_layout_details_level_1.DocumentLayoutDetailsLevel.Row) + lp.row.getLastBox().getEndPosition())
                    return startFieldPos;
                else
                    break;
            }
        }
        return pos;
    };
    return FloatingObjectDragDropChangePositionCommand;
}(command_base_1.CommandBase));
exports.FloatingObjectDragDropChangePositionCommand = FloatingObjectDragDropChangePositionCommand;
var FloatingObjectDragDropChangePositionCommandParameters = (function (_super) {
    tslib_1.__extends(FloatingObjectDragDropChangePositionCommandParameters, _super);
    function FloatingObjectDragDropChangePositionCommandParameters(control, startPageIndex, endPageIndex, finalPoint, finalClickPoint) {
        var _this = _super.call(this, control) || this;
        _this.startPageIndex = startPageIndex;
        _this.endPageIndex = endPageIndex;
        _this.finalPoint = finalPoint;
        _this.finalClickPoint = finalClickPoint;
        return _this;
    }
    return FloatingObjectDragDropChangePositionCommandParameters;
}(command_base_1.CommandOptions));
exports.FloatingObjectDragDropChangePositionCommandParameters = FloatingObjectDragDropChangePositionCommandParameters;
