// devexpress-richedit/lib/core/model/manipulators/text-manipulator/text-manipulator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var fixed_1 = require("@devexpress/utils/lib/intervals/fixed");
var simple_run_inserted_1 = require("../../changes/sub-document/text/simple-run-inserted");
var history_item_1 = require("../../history/base/history-item");
var insert_text_history_item_1 = require("../../history/items/insert-text-history-item");
var runs_base_manipulator_1 = require("../runs-base-manipulator");
var chunk_size_corrector_1 = require("./chunk-size-corrector");
var InsertTextViaHistoryResult = (function () {
    function InsertTextViaHistoryResult(inserted, insertedInterval) {
        this.inserted = inserted;
        this.insertedInterval = insertedInterval;
    }
    return InsertTextViaHistoryResult;
}());
exports.InsertTextViaHistoryResult = InsertTextViaHistoryResult;
var TextManipulator = (function (_super) {
    tslib_1.__extends(TextManipulator, _super);
    function TextManipulator(manipulator) {
        var _this = _super.call(this, manipulator) || this;
        _this.chunkSizeCorrector = new chunk_size_corrector_1.ChunkSizeCorrector();
        return _this;
    }
    TextManipulator.prototype.insertTextViaHistory = function (params) {
        if (!params.correctAndCheckParams())
            return new InsertTextViaHistoryResult(false, null);
        this.history.addAndRedo(new insert_text_history_item_1.InsertTextHistoryItem(this.modelManipulator, params));
        return new InsertTextViaHistoryResult(true, new fixed_1.FixedInterval(params.subDocPos.position, params.text.length));
    };
    TextManipulator.prototype.insertTextInner = function (params) {
        var insertedRun = this.insertRunInternal(params.subDocPos, params.charPropsBundle, params.runType, params.text);
        var textRun = params.subDocPos.subDocument.chunks[insertedRun.chunkIndex].textRuns[insertedRun.runIndex];
        textRun.paragraph.length += params.text.length;
        this.chunkSizeCorrector.correctChunkSizeAtInsertPosition(params.subDocPos.subDocument, params.subDocPos.position);
        this.modelManipulator.notifyModelChanged(new simple_run_inserted_1.SimpleRunInsertedSubDocumentChange(params.subDocPos.subDocument.id, params.subDocPos.position, params.text.length, textRun.maskedCharacterProperties, textRun.characterStyle, params.runType, params.text));
    };
    TextManipulator.prototype.getLastModifiableHistoryItem = function (checkType) {
        var history = this.history;
        if (!history)
            return null;
        var historyItems = history.historyItems;
        return history.currentIndex == historyItems.length - 1 ? this.getPrevItem(historyItems, checkType).item : null;
    };
    TextManipulator.prototype.getPrevItem = function (historyItems, checkType) {
        for (var ind = historyItems.length - 1, item = void 0; item = historyItems[ind]; ind--) {
            if (!item.canBeMerged())
                return new PrevItemResult(true, null);
            if (checkType(item))
                return new PrevItemResult(false, item);
            if (item instanceof history_item_1.CompositionHistoryItem) {
                var lowLevelResult = this.getPrevItem(item.historyItems, checkType);
                if (lowLevelResult.cantBeMerged || !lowLevelResult.cantBeMerged && lowLevelResult.item)
                    return lowLevelResult;
            }
        }
        return new PrevItemResult(false, null);
    };
    return TextManipulator;
}(runs_base_manipulator_1.RunsBaseManipulator));
exports.TextManipulator = TextManipulator;
var PrevItemResult = (function () {
    function PrevItemResult(cantBeMerged, item) {
        this.cantBeMerged = cantBeMerged;
        this.item = item;
    }
    return PrevItemResult;
}());
